### PR TITLE
AE-484 USDC Balance Change Monitor

### PR DESCRIPTION
**Pull request by arbitrary-kdawg**

### CODE REVIEW

File: contract-addresses.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "Vault": "0x20c2554429dD269FB7785a1FCA08a04a70484E5a",
```
Comment (echoAE) [inline contract-addresses.json:2]: In the current structure of this repository, these addresses should live in `protocol-data.json`.  These two addresses are from version `v0.11.0` (and `v0.11.1`) of the `@perp/lushan` npm package.

We should nail down whether we will use the `@perp/lushan` npm package to directly get these addresses and what version to use if we use it.  If we do not use it, we should determine how to extract and store the appropriate contract/account addresses within the repository.

File: contract-addresses.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+  "InsuranceFund": "0xe7f52b4D2DB7211A93D7C864b2B7205F79F2Ee10"
```
Comment (echoAE) [inline contract-addresses.json:3]: In the current structure of this repository, these addresses should live in `protocol-data.json`.  These two addresses are from version `v0.11.0` (and `v0.11.1`) of the `@perp/lushan` npm package.

We should nail down whether we will use the `@perp/lushan` npm package to directly get these addresses and what version to use if we use it.  If we do not use it, we should determine how to extract and store the appropriate contract/account addresses within the repository.

File: package.json [hunk 3]
```diff
@@ -15,6 +15,7 @@
   },
   "dependencies": {
     "@uniswap/v3-core": "^1.0.0",
+    "@perp/lushan": "^0.11.1",
```
Comment (echoAE) [inline package.json:18]: This package does not appear to be used in this handler.

File: src/usdc-balance-change/usdc-balance-change.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+const initializeData = {};
+
+// load account addresses to monitor
+const accountAddresses = require('../../account-addresses.json');
+
+// load contract addresses to monitor
+const contractAddresses = require('../../contract-addresses.json');
+
+// provide ABI for USDC balanceOf()
+const usdcAbi = [
```
Comment (echoAE) [inline src/usdc-balance-change/usdc-balance-change.js:19]: Is there an ABI in a JSON file that could be copied into `forta-perpfi-suite/abis/`?

Alternatively, is there already an ABI there that matches the ABI for this (USDC)?
`BaseToken.json`?

File: src/usdc-balance-change/usdc-balance-change.spec.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+    it('when all balance changes are below the threshold', async () => {
+      // simulate a changing balance under the threshold
+      // 5.0, 5.1, 5.2, 5.3, 5.4 USDC
+      mockUsdcContract.balanceOf = jest.fn()
+        .mockResolvedValueOnce(ethers.BigNumber.from(5000000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(5100000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(5200000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(5300000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(5400000));
```
Comment (echoAE) [inline src/usdc-balance-change/usdc-balance-change.spec.js:87]: Minor point: Create a variable to store `5400000` to use here and in the `expect` on line 102.

File: src/usdc-balance-change/usdc-balance-change.spec.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+  describe('reports', () => {
+    it('when the balance change is above the threshold and within the block window', async () => {
+      // simulate a changing balance over the threshold
+      // 3.0, 3.1, 3.2, 3.3, 3.6 USDC
+      mockUsdcContract.balanceOf = jest.fn()
+        .mockResolvedValueOnce(ethers.BigNumber.from(3000000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(3100000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(3200000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(3300000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(3600000));
```
Comment (echoAE) [inline src/usdc-balance-change/usdc-balance-change.spec.js:148]: Same comment as the previous two tests: Create a variable to store the `3600000` value and use it here and in the `expect` statement on line 178.

**Review by echoAE — CHANGES_REQUESTED** [review]
We should have a brief discussion about the origin of contract/account addresses and the use of protocol npm packages.

There is also some code in another branch with a `common.js` file that facilitates extraction of addresses and abis from the protocol JSON files.  Integrating the use of that module with handlers could be the subject of another ticket.

File: package.json (hunk 3 above)
Comment (arbitrary-kdawg) [reply package.json:18]: Will move to devDependencies in package.json

File: agent-config.json [hunk 7]
```diff
@@ -58,5 +58,10 @@
   "pendingTransactions": {
     "timeWindowSeconds": 60,
     "txThreshold": 10
+  },
+  "usdcBalanceChange": {
+    "USDC_ADDRESS" : "0xeb8f08a975ab53e34d8a0330e0d34de942c95926",
```
[the remark this answers is not in the document]
Comment (arbitrary-kdawg) [reply agent-config.json:63]: Will change to 0x40DFcFEB94575Aaf4a97EBE020850e13f84a46E5.  All the addresses are changing and we will need a final review with the Perp.Fi team.

File: Dockerfile [hunk 8]
```diff
@@ -14,6 +14,7 @@ WORKDIR /app
 COPY ./src ./src
 COPY agent-config.json ./
 COPY account-addresses.json ./
+COPY contract-addresses.json ./
```
[the remark this answers is not in the document]
Comment (arbitrary-kdawg) [reply Dockerfile:17]: Discussed offline - contract-addresses.json will be eliminated.

File: contract-addresses.json (hunk 2 above)
Comment (arbitrary-kdawg) [reply contract-addresses.json:3]: Discussed offline - contract-addresses.json will be eliminated.

File: src/usdc-balance-change/usdc-balance-change.spec.js (hunk 5 above)
Comment (arbitrary-kdawg) [reply src/usdc-balance-change/usdc-balance-change.spec.js:87]: Discussed offline; will leave as-is

File: src/usdc-balance-change/usdc-balance-change.spec.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      // 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7 USDC
+      mockUsdcContract.balanceOf = jest.fn()
+        .mockResolvedValueOnce(ethers.BigNumber.from(6000000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(6100000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(6200000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(6300000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(6400000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(6500000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(6600000))
+        .mockResolvedValueOnce(ethers.BigNumber.from(6700000));
```
[the remark this answers is not in the document]
Comment (arbitrary-kdawg) [reply src/usdc-balance-change/usdc-balance-change.spec.js:121]: Discussed offline; will leave as-is

File: src/usdc-balance-change/usdc-balance-change.spec.js (hunk 6 above)
Comment (arbitrary-kdawg) [reply src/usdc-balance-change/usdc-balance-change.spec.js:148]: Discussed offline; will leave as-is

File: src/usdc-balance-change/usdc-balance-change.js [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+      }
+
+      // once we see blockWindow+1 blocks, we have enough data to check for a balance change
+      //
+      // e.g. to check two balances ~1 minute apart, assuming a block interval of ~15 seconds,
+      //      we need 5 samples:
+      //
+      //      balanceHistory = [BAL1, BAL2, BAL3, BAL4, BAL5]
+      //
+      //      The time between when BAL1 and BAL5 was recorded will be approximately 1 minute
```
[the remark this answers is not in the document]
Comment (arbitrary-kdawg) [reply src/usdc-balance-change/usdc-balance-change.js:98]: Fixed

File: src/usdc-balance-change/usdc-balance-change.js (hunk 4 above)
Comment (arbitrary-kdawg) [reply src/usdc-balance-change/usdc-balance-change.js:19]: Switched to using IERC20Metadata.json from @perp/lushan, which provides the balanceOf() method.

File: src/usdc-balance-change/usdc-balance-change.js [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+      } catch (err) {
+        console.error(err);
+      } finally {
+        // if balanceOf() threw an exception, balance will be undefined
+        addresses[address].balanceHistory.push(balance);
+      }
+
+      // once we see blockWindow+1 blocks, we have enough data to check for a balance change
+      //
+      // e.g. to check two balances ~1 minute apart, assuming a block interval of ~15 seconds,
```
[the remark this answers is not in the document]
Comment (arbitrary-kdawg) [reply src/usdc-balance-change/usdc-balance-change.js:90]: I went ahead and implemented the change to detect more "noise" in the balance changes--it is trivial to revert if this is not the desired approach.

**Review by echoAE — APPROVED** [review]

**Review by arbitrary-sparrow — APPROVED** [review]

**Pull request merged**